lib/Listener/NewMessageClassificationListener.php

```diff
@@ -67,11 +74,18 @@ public function handle(Event $event): void {
 			}
 		}
 
-		// if the message is already flagged as important, we won't classify it again.
-		$messages = array_filter($event->getMessages(), function ($message) {
-			return ($message->getFlagImportant() === false);
+		$messages = $event->getMessages();
+
+		// if this is a message that's been flagged / tagged as important before, we don't want to reclassify it again.
+		$doNotReclassify = $this->tagMapper->getTaggedMessageIdsForMessages($event->getMessages(), $event->getAccount()->getUserId(), Tag::LABEL_IMPORTANT);
+		$messages = array_filter($messages, static function ($message) use ($doNotReclassify) {
+			return ($message->getFlagImportant() === false || in_array($message->getMessageId(), $doNotReclassify, true));
 		});
 
+		// we can manually build the tag here, because we do the DB lookup in the tagMessage method
```

Comment:
related to the comment above I'm not a fan of this. Making assumptions about the implementation over there is prone to bugs in the future as code is changed for other features.

Can we just load `$label1` tag instance here and assign the actual entity instance that has the correct ID?

Reply:
Sure thing, but that would use the `getTagByImapLabel` method again, which is indeed a bottleneck. We do need to check if the tag exists, which means checking every tag on every message against the DB to either insert it or set the right `id` - because if we don't we get an Unique Constraint violation on the `imap_label - user_id` combination IIRC.

I'll open another PR for the bulkUpdate / insert method to speed it up.